src/app/services/chart.service.ts

```diff
@@ -0,0 +1,111 @@
+import { HttpClient } from "@angular/common/http";
+import { AppComponent } from "../app.component";
+import { DatePipe } from "@angular/common";
+import { Injectable } from '@angular/core';
+import { SignInService } from "./sign-in.service";
+
+export interface ChartSpec {
+  chartType: string;
+  options: {};
+  dataTable: any;
+  error: any;
+}
+
+let makeChartSpec = (chartType, options): ChartSpec => {
+  return {
+    chartType,
+    options,
+    dataTable: null,
+    error: null
+  };
+};
+
+let identity = x => x;
+
+@Injectable()
+export class ChartService {
+  constructor(
+    private http: HttpClient,
+    private app: AppComponent,
+    private datePipe: DatePipe,
+    private signIn: SignInService
+  ) {}
+
+  fetchChart(url: string, chartSpec: ChartSpec, mapData = identity) {
+    return this.http.get(url).map(data => {
+      console.log(data);
+      if (Object.keys(data).length > 0) {
+        chartSpec.dataTable = mapData(data);
```

Comment:
this `mapData` function is here because the histogram chart contains its data inside a nested structure that needs to be unwrapped.